File: datastructure/sparsetable.cpp
#include <bits/stdc++.h>

using namespace std;
#define MAX 500

int lookup[MAX][MAX];

void buildSparseTable(int arr[], int n)
{
    for(int i = 0; i<n; i++)
    {
        lookup[i][0] = arr[i];
    }

    //compute values from smaller to bigger intervals
    for(int j =1; (1 << j) <=n; j++)
    {
        //compute minimum value for all intervals with size 2^j

        for(int i=0; (i + (1 << j) -1) <n; i++)
        {
            if(lookup[i][j-1] < lookup[i + (1 << (j-1))][j-1])
                {
                    lookup[i][j] = lookup[i][j-1];
                }
            else
            {
                lookup[i][j] = lookup[i + (1 << (j-1))][j-1];
            }
        }
    }
}

// returns minimum of arr[L..R]
int query(int L, int R)
{
    //find hghest power of 2 that is smaller
    //than or equal to count of elements in given range. For [2,10], j= 3

    int j = (int) log2(R-L+1);

    //compute minimum of last 2^j elements with first 2^j elements in range
    // for [2,10], we compare arr[lookup[0][3]] and arr[lookup[3][3]],

    if(lookup[L][j] <= lookup[R -(1 << j) +1][j])
    {
        return lookup[L][j];
    } 
    else{
        return lookup[R -(1 << j) +1][j];
    }
}


int main()
{
    int a[] = {7,2,3,0,5,10,3,12,18};
    int n = sizeof(a) / sizeof(a[0]);
    buildSparseTable(a,n);

    cout << query(4,7) << endl;

    return 0;
}

File: datastructure/sparse_table.py
import math 

def buildSparseTable(arr, n, lookup):
    for i in range(0,n):
        lookup[i][0]  = arr[i]
    
    j = 1

    while 2**j < n:
        i = 0

        while (i + (1 << j) -1) < n:

            if lookup[i][j-1] < lookup[i + (2**(j-1))][j-1]:
                lookup[i][j] = lookup[i][j-1]
            else:
                lookup[i][j] = lookup[i + (2 **(j-1))][j-1]
            i +=1
        j +=1


def query(L,R):
    j = int(math.log2(R-L + 1))

    if lookup[L][j] <= lookup[R - 2**j + 1][j]:
        return lookup[L][j]
    else:
        return lookup[R - (2**j ) + 1][j]
a = [7, 2, 3, 0, 5, 10, 3, 12, 18]
n = len(a)
MAX = n
     
    # lookup[i][j] is going to store minimum
    # value in arr[i..j]. Ideally lookup table
    # size should not be fixed and should be
    # determined using n Log n. It is kept
    # constant to keep code simple.
lookup = [[0 for i in range(MAX)]
                 for j in range(MAX)]
buildSparseTable(a,n,lookup)

print(lookup)

print(query(0, 4))
print(query(4, 7))
print(query(7, 8))